fc_util/src/arg_parser.rs

```diff
@@ -0,0 +1,581 @@
+pub struct ArgInfo<'a> {
+    name: &'a str,
+    required: bool,
+    requires: Option<&'a str>,
+    takes_value: bool,
+    default_value: Option<&'a str>,
+    optional: bool,
+    help: Option<&'a str>,
+    user_value: Option<String>,
+}
+
+impl<'a> ArgInfo<'a> {
+    pub fn new(name: &'a str) -> ArgInfo<'a> {
+        ArgInfo {
+            name,
+            required: false,
+            requires: None,
+            takes_value: false,
+            default_value: None,
+            optional: false,
+            help: None,
+            user_value: None,
+        }
+    }
+
+    pub fn required(mut self, required: bool) -> ArgInfo<'a> {
+        self.required = required;
+        self
+    }
+
+    pub fn requires(mut self, requires: &'a str) -> ArgInfo<'a> {
+        self.requires = Some(requires);
+        self
+    }
+
+    pub fn takes_value(mut self, takes_value: bool) -> ArgInfo<'a> {
+        self.takes_value = takes_value;
+        self
+    }
+
+    pub fn default_value(mut self, default_value: &'a str) -> ArgInfo<'a> {
+        self.default_value = Some(default_value);
+        self
+    }
+
+    pub fn optional(mut self, optional: bool) -> ArgInfo<'a> {
+        self.optional = optional;
+        self
+    }
+
+    pub fn help(mut self, help: &'a str) -> ArgInfo<'a> {
+        self.help = Some(help);
+        self
+    }
+
+    pub fn update_value(mut self, value: String) -> ArgInfo<'a> {
+        self.user_value = Some(value);
+        self
+    }
+
+    pub fn format_help(&self) -> String {
+        if let Some(help) = self.help {
+            if self.takes_value {
+                format!("--{} <{}>: {}", self.name, self.name, help)
+            } else {
+                format!("--{}: {}", self.name, help)
+            }
+        } else if self.takes_value {
+            format!("--{} <{}>", self.name, self.name)
+        } else {
+            format!("--{}", self.name)
+        }
+    }
+}
+
+impl<'a> PartialEq for ArgInfo<'a> {
+    fn eq(&self, other: &Self) -> bool {
+        self.name == other.name
+    }
+}
+
+impl<'a> Hash for ArgInfo<'a> {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.name.hash(state);
+    }
+}
+
+impl<'a> Borrow<str> for ArgInfo<'a> {
+    fn borrow(&self) -> &str {
+        self.name
+    }
+}
+
+#[derive(Debug, Default, PartialEq)]
+pub struct ArgParser<'a> {
+    args_matches: HashSet<ArgInfo<'a>>,
+    extra_args: Vec<String>,
+}
+
+impl<'a> ArgParser<'a> {
+    pub fn arg(mut self, arg_info: ArgInfo<'a>) -> Self {
+        self.args_matches.insert(arg_info);
+        self
+    }
+
+    pub fn value(&self, arg_name: &'static str) -> Option<String> {
+        self.args_matches.get(arg_name).unwrap().user_value.clone()
+    }
+
+    pub fn extra_args(self) -> Vec<String> {
+        self.extra_args
+    }
+
+    // Used for a general validation of the arguments provided by user.
+    fn is_valid(&self, argument_info: &ArgInfo<'a>, index: usize, args: &[String]) -> Result<()> {
```

Comment:
I wouldn't send the entire command line array to this method (`args: &[String]`) because it doesn't seem to be needed. I would only send the value if it exists (`maybe_value: Option<&String>`).